packages/@aws-cdk/example-construct-library/lib/example-resource.ts

```diff
@@ -0,0 +1,501 @@
+ *     Those resources are represented by the class with the name identical to the resource -
+ *     {@link ExampleResource} in our case, which implements {@link IExampleResource}.
+ *   2. It can be a resource that exists already, and is not managed by the CDK code,
+ *     but needs to be referenced in your infrastructure definition code.
+ *     Those kinds of instances are returned from static `fromXyz(Name/Arn/Attributes)` methods -
+ *     in our case, the {@link ExampleResource.fromExampleResourceName} method.
+ *     In general, those kinds of resources do not allow any sort of mutating operations to be performed on them
+ *     (the exception is when they can be changed by creating a different resource -
+ *     IAM Roles, which you can attach multiple IAM Policies to,
+ *     are the canonical example of this sort of resource),
+ *     as they are not part of the CloudFormation stack that is created by the CDK.
+ *
+ *  So, an interface like {@link IExampleResource} represents a resource that *might* be mutable,
+ *  while the {@link ExampleResource} class represents a resource that definitely is mutable.
+ *  In almost all cases, when you're using this resource as the type of a property in a different resource definition,
+ *  you want to use {@link IExampleResource} as the type, not {@link ExampleResource}.
+ *
+ *  The interface for the resource should have at least 2 (readonly) properties
+ *  that represent the ARN and the physical name of the resource -
+ *  in our example, those are {@link exampleResourceArn} and {@link exampleResourceName}.
+ *
+ *  The interface defines the behaviors the resource exhibits.
+ *  Common behaviors are:
+ *    - {@link addToRolePolicy} for resources that are tied to an IAM Role
+ *    - grantXyz() methods (represented by {@link grantRead} in this example)
+ *    - onXyz() CloudWatch Events methods (represented by {@link onEvent} in this example)
+ *    - metricXyz() CloudWatch Metric methods (represented by {@link metricSth} in this example)
+ *
+ *  Of course, other behaviors are possible -
+ *  it all depends on the capabilities of the underlying resource that is being modeled.
+ *
+ *  This interface must always extend the IResource interface from the core module.
+ *  It can also extend some other common interfaces that add various default behaviors -
+ *  some examples are shown below.
+ */
+export interface IExampleResource extends
+  // all L2 interfaces need to extend IResource
+  core.IResource,
+
+  // Only for resources that have an associated IAM Role.
+  // Allows this resource to be the target in calls like bucket.grantRead(exampleResource).
+  iam.IGrantable,
+
+  // only for resources that are in a VPC and have SecurityGroups controlling their traffic
+  ec2.IConnectable {
+
+  /**
+   * The ARN of example resource.
+   * Equivalent to doing `{ 'Fn::GetAtt': ['LogicalId', 'Arn' ]}`
+   * in CloudFormation.
+   * It needs to be annotated with '@attribute'.
+   *
+   * @attribute
```

Comment:
Only if the ARN is a return value of the CloudFormation resource.

Reply:
Sorry, I don't understand this comment...?

Reply:
`@attribute` should be added only for properties that map to return values (via `Ref` or `GetAtt`) from CloudFormation. Not all resource types return ARN as a property.

Reply:
Added explanation.